exonum-java-binding/testkit/src/test/java/com/exonum/binding/testkit/TestKitExtensionTest.java

```diff
@@ -0,0 +1,205 @@
+  }
+
+  private void checkFailedEvents(Events testEvents, String expectedMessage) {
+    Events failedEvents = testEvents.failed();
+    assertThat(failedEvents.count()).isEqualTo(1);
+    String exceptionMessage = getFailedEventExceptionMessage(failedEvents.list().get(0));
+    assertThat(exceptionMessage).contains(expectedMessage);
+  }
+
+  private String getFailedEventExceptionMessage(Event event) {
+    TestExecutionResult testExecutionResult = event.getPayload(TestExecutionResult.class).get();
+    Throwable throwable = testExecutionResult.getThrowable().get();
+    return throwable.getMessage();
+  }
+
+  private Events getTestCaseEvents(Class<?> testCaseClass) {
+    return getTestCaseEngineExecutionResults(testCaseClass)
+        .tests();
+  }
+
+  private Events getAllTestCaseEvents(Class<?> testCaseClass) {
+    return getTestCaseEngineExecutionResults(testCaseClass)
+        .all();
+  }
+
+  private EngineExecutionResults getTestCaseEngineExecutionResults(Class<?> testCaseClass) {
+    return EngineTestKit.engine("junit-jupiter")
+        .selectors(selectClass(testCaseClass))
+        .execute();
+  }
+
+  static class TestKitInstantiationTestCase {
+
+    @RegisterExtension
+    static TestKitExtension testKitExtension = new TestKitExtension(defaultBuilder);
+
+    static TestKit instantiatedTestKit;
+
+    @BeforeEach
+    void beforeEachTestKitInject(TestKit testKit) {
+      // Creates TestKit based on default builder
+      instantiatedTestKit = testKit;
+
+      // Check that TestKit was instantiated with a correct service
+      TestService service = testKit.getService(TestService.SERVICE_ID, TestService.class);
+      assertThat(service.getName()).isEqualTo(TestService.SERVICE_NAME);
+
+      // Check that main TestKit node is a validator
+      assertThat(testKit.getEmulatedNode().getValidatorId()).isNotEmpty();
+    }
+
+    @Test
+    void testDefaultTestKit(TestKit testKit) {
+      assertThat(testKit).isEqualTo(instantiatedTestKit);
+    }
+
+    @AfterEach
+    void afterEachTestKitInject(TestKit testKit) {
+      assertThat(testKit).isEqualTo(instantiatedTestKit);
+    }
+  }
+
+  static class BeforeEachInstantiationTestCase {
+
+    @RegisterExtension
+    static TestKitExtension testKitExtension = new TestKitExtension(defaultBuilder);
+
+    @BeforeEach
+    void beforeEach(TestKit testKit) {
+      // Should pass
+    }
+
+    @Test
+    void testTestKitReconfiguration(@Auditor TestKit testKit) {
+      // Shouldn't be executed
+    }
+  }
+
+  static class AfterEachInstantiationTestCase {
+
+    @RegisterExtension
+    static TestKitExtension testKitExtension = new TestKitExtension(defaultBuilder);
+
+    @AfterEach
+    void afterEach(@ValidatorCount(validatorCount = 8) TestKit testKit) {
+      // Shouldn't be executed
+    }
+
+    @Test
+    void testTestKitReconfiguration(TestKit testKit) {
+      // This test should pass
+    }
+  }
+
+  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
```

Comment:
Is it needed if the extension is already in `static` field?

Reply:
Yes, still needed

Reply:
But why?

> Static Fields
If a @RegisterExtension field is static, the extension will be registered after extensions that are registered at the class level via `@ExtendWith`. Such static extensions are not limited in which extension APIs they can implement. **Extensions registered via static fields may therefore implement class-level and instance-level extension APIs such as BeforeAllCallback, AfterAllCallback,** and TestInstancePostProcessor as well as method-level extension APIs such as BeforeEachCallback, etc.

Reply:
❓
On top of that — the extension is no longer static.

Reply:
❓
On top of that — the extension is no longer static.

Reply:
Sorry, missed your previous question. If we had both an extension and `@BeforeAll` method `static`, then it's all good - we don't need an annotation. We had `@BeforeAll` not static though when you originally commented on that, that's why we had `@TestInstance(TestInstance.Lifecycle.PER_CLASS)` instead.

Now as extension is not `static`, we need this annotation, otherwise we get `"@BeforeAll method <method> must be static unless the test class is annotated with @TestInstance(Lifecycle.PER_CLASS)."`.

Reply:
Ok, let's keep them non-static.